extensions/git/src/terminal.ts

```diff
@@ -10,32 +10,31 @@ export class TerminalEnvironmentManager {
 
 	private readonly disposable: IDisposable;
 
-	private _enabled = false;
-	private set enabled(enabled: boolean) {
-		if (this._enabled === enabled) {
-			return;
-		}
-
-		this._enabled = enabled;
-		this.context.environmentVariableCollection.clear();
-
-		if (enabled) {
-			for (const name of Object.keys(this.env)) {
-				this.context.environmentVariableCollection.replace(name, this.env[name]);
-			}
-		}
-	}
-
 	constructor(private readonly context: ExtensionContext, private readonly env: { [key: string]: string }) {
-		this.disposable = filterEvent(workspace.onDidChangeConfiguration, e => e.affectsConfiguration('git'))
-			(this.refresh, this);
+		this.disposable = filterEvent(workspace.onDidChangeConfiguration, e =>
+			e.affectsConfiguration('git.enabled') ||
+			e.affectsConfiguration('git.terminalAuthentication') ||
+			e.affectsConfiguration('git.terminalGitEditor')
+		)(this.refresh, this);
 
 		this.refresh();
 	}
 
 	private refresh(): void {
+		this.context.environmentVariableCollection.clear();
 		const config = workspace.getConfiguration('git', null);
-		this.enabled = config.get<boolean>('enabled', true) && config.get('terminalAuthentication', true);
+
+		if (!config.get<boolean>('enabled', true)) {
+			return;
+		}
+
+		for (const name of Object.keys(this.env)) {
+			if ((name.toUpperCase() !== 'GIT_ASKPASS' && name.toUpperCase() !== 'GIT_EDITOR') ||
+				(name.toUpperCase() === 'GIT_ASKPASS' && config.get<boolean>('terminalAuthentication', true)) ||
+				(name.toUpperCase() === 'GIT_EDITOR' && config.get<boolean>('terminalGitEditor', true))) {
+				this.context.environmentVariableCollection.replace(name, this.env[name]);
```

Comment:
Not too happy that the TerminalEnvironmentManager now knows about very specific stuff. Maybe it's time to introduce a concept of TerminalEnvironmentProvider and have Askpass and GitEditor be such providers.

Reply:
@joaomoreno, pushed another commit to clean this up. Let me know what you think. Thanks!